enhancements/machine-config/securing-mcs.md

```diff
@@ -0,0 +1,121 @@
+
+[^1]: "This is specifically because the blocking rules were done as a quick hack. In theory, we could block every pod from accessing only the MCS without blocking anything else, but that would require noticing when masters were added/moved and updating the rules for every pod when that happened. Since this was supposed to have been a temporary hack, we didn't bother, and instead just blocked access to ports 22623 and 22624 on all IPs." -@danwinship
+
+## Motivation
+
+This blanket network policy prevents traffic destined for ports 22623 and 22624, regardless of the source and destination IP addresses. This has the potential to impact customer workloads, but also internal initiatives (e.g. OpenShift on OpenShift is complicated by this restriction).
+
+### Goals
+
+- Remove the port restriction on 22623 and 22624
+- Ensure that the node bootstrap flow remains secure
+
+### Non-Goals
+
+- Allow customers to use Machine Configs to provide pre-bootstrap configuration
+
+## Proposal
+
+Security assumptions:
+
+  - The CSR bootstrap token is not considered sensitive and being accessible to all applications in the cluster is acceptable.
+  - The pull secret is not considered sensitive and being accessible to all applications in the cluster is acceptable.
+  - The contents of Machine Configs are potentially sensitive, especially if the Machine Configs are customer-provided.
+
+If it can be ensured that no one is able to access sensitive data through the Machine Config Server, the port restrictions can be safely removed. In order to achieve this, Machine Configs can be split into two groups: pre-node-approved configs and post-node-approved configs. The pre-node-approved configs are ones which do not contain sensitive information and contain just enough to configure new nodes to submit a certificate signing request (CSR) to the API server. The post-node-approved configs contain everything else, including the overwhelming majority of customer-provided Machine Configs. When a new machine is created, it fetches its Ignition config from the Machine Config Server, which only serves the pre-node-approved configs. At this point, the machine has enough information to submit a CSR, but nothing sensitive (note that for the purposes of this proposal, the CSR bootstrap token and pull secret are not considered sensitive). Once the CSR has been approved and a kubeconfig returned, the Machine Config Daemon is able to directly enumerate Machine Configs, including the post-node-approved ones, so it then applies any remaining configuration to the machine and reboots if necessary. At this point, the machine is fully configured.
```

Comment:
Today we carry effectively 1.63 implementations of Ignition - one in the upstream project that runs in the initramfs, and  63% of Ignition in the MCD that runs in the real root.  If do this, then we'd hit all limitations of the 63% MCD-Ignition implementation upfront.  For example, while this isn't important to OpenShift, today the MCO doesn't support adding/changing users.

We may hit edge cases around e.g. parent directory creation, file write ordering etc.  Those are all bugs that *should* be fixed of course, but we also tend to not hit them today because use of "day 2" MCs is so limited.

If we go this route it may be worth a spike around how we can have the MCO actually reuse Ignition upstream code directly for this.